File: python/leetcode/256.py
class Solution:
    def minCost(self, costs):
        if not costs: return 0
        pre = [0] * 3
        for now in costs:
            pre = [now[i] + min(pre[:i], pre[i + 1:]) for i in range(3)]
        return min(pre)

class Solution:
    def minCost(self, costs):
        if not costs: return 0
        pre = costs[0][:]
        now = [0] * 3
        for i in range(1, n):
            now[0] = costs[i][0] + min(pre[1] + pre[2])
            now[1] = costs[i][1] + min(pre[0] + pre[2])
            now[2] = costs[i][2] + min(pre[0] + pre[1])
            pre = now
        return min(pre)

File: cpp/leetcode/256.cpp
class Solution {
public:
    int minCost(vector<vector<int>>& costs) {
        if(costs.empty() || costs[0].empty()) return 0;
        int n = costs.size();
        vector<int> pre = {costs[0][0], costs[0][1], costs[0][2]};
        vector<int> cur(3, 0);
        for(int i = 1; i < n; i++){
            cur[0] = min(pre[1], pre[2]) + costs[i][0];
            cur[1] = min(pre[0], pre[2]) + costs[i][1];
            cur[2] = min(pre[0], pre[1]) + costs[i][2];
            pre = cur;
        }
        return min(pre[0], min(pre[1], pre[2]));
    }
};